### PR TITLE
Allow Dynamic Service Binding

### PR DESCRIPTION
**Pull request by pete-tempo**
Allows for dynamic service binding, removing the need to restart the scripting server during the lifecycle of an editor or engine session. This is accomplished by discovering the CDO of all classes implementing ITempoScriptable, and gathering their services and rpcs before server initialization, and then allowing their handlers to be bound later (in a step now called "Activating" the services, from non-CDO instances of those classes. This unlocks RPCs that span multiple world lifetimes, for example LoadLevel, which starts loading a new level and only returns once the new level has finished loading.

### CODE REVIEW

File: TempoCore/Source/TempoScripting/Public/TempoScriptingServer.h [hunk 1]
```diff
@@ -43,61 +51,85 @@ struct TRequestHandler
...
+	void Activate(UserObjectType* UserObject)
 	{
-		checkf(!HandleDelegate.IsBound(), TEXT("Handle delegate is already bound. Multiple handlers is not supported."));
-		HandleDelegate.BindRaw(UserObject, HandleFunction, Forward<VarTypes>(Vars)...);
-		return *this;
+		if (!ensureAlwaysMsgf(!UserObject->HasAnyFlags(RF_ClassDefaultObject), TEXT("Bound called on CDO! Do not call BindObjectToService from RegisterScriptingServices.")))
+		{
+			return;
+		}
+		if (HandleDelegate.IsBound() && BindType == EBindType::Lambda)
```
[the remark this answers is not in the document]
Comment (pete-tempo) [reply TempoCore/Source/TempoScripting/Public/TempoScriptingServer.h:66]: I think it's good the way it is. Capturing a UObject in a lambda is a programming error.

File: TempoCore/Source/TempoScripting/Private/TempoScriptingServer.cpp [hunk 2]
```diff
@@ -28,7 +27,6 @@ FTempoScriptingServer::FTempoScriptingServer()
 	{
 		if (UTempoCoreUtils::IsGameWorld(World))
 		{
-			Reinitialize();
```
[the remark this answers is not in the document]
Comment (pete-tempo) [reply TempoCore/Source/TempoScripting/Private/TempoScriptingServer.cpp:31]: We don't even need to Reinitialize after hot reload anymore. It all works. However, we still have to Reinitialize when settings are changed, like the server port, and in that case we have to take care to re-activate any servers that were activated. Added that.

File: TempoCore/TempoCore.uplugin [hunk 3]
```diff
@@ -28,7 +28,7 @@
 		{
 			"Name": "TempoScripting",
 			"Type": "Runtime",
-			"LoadingPhase": "Default"
+			"LoadingPhase": "PostDefault"
```
[the remark this answers is not in the document]
Comment (pete-tempo) [reply TempoCore/TempoCore.uplugin]: No need anymore

**Review by gio-tempo — APPROVED** [review]

File: TempoCore/Source/TempoScripting/Private/TempoScriptingServer.cpp [hunk 4]
```diff
@@ -19,14 +19,19 @@ FTempoScriptingServer::FTempoScriptingServer()
...
+				}));
 		}
 	});
 	OnPreWorldFinishDestroyHandle = FWorldDelegates::OnPreWorldFinishDestroy.AddLambda(
 		[this](UWorld* World)
 	{
 		if (UTempoCoreUtils::IsGameWorld(World))
 		{
+			// Tick one more time to give deactivated services a chance to flush their final messages
+			TickInternal();
```
Comment (pete-tempo) [inline TempoCore/Source/TempoScripting/Private/TempoScriptingServer.cpp:34]: Need to do this on game shutdown too?

File: TempoCore/Source/TempoCore/Private/TempoCoreServiceSubsystem.cpp [hunk 5]
```diff
@@ -0,0 +1,100 @@
+// Copyright Tempo Simulation, LLC. All Rights Reserved
```
Comment (pete-tempo) [inline TempoCore/Source/TempoCore/Private/TempoCoreServiceSubsystem.cpp:1]: Remove this?

File: TempoCore/Source/TempoScripting/Private/TempoScriptingServer.cpp (hunk 4 above)
Comment (pete-tempo) [reply TempoCore/Source/TempoScripting/Private/TempoScriptingServer.cpp:34]: No, it's not necessary. This one is sufficient.

**Review by gio-tempo — APPROVED** [review]

**Pull request merged**